### DISCUSSION TITLE
[gym] Use Xcode 13's cloud signing

### DISCUSSION THREAD

**Discussion by tahirmt** [fastlane/fastlane#19973] [Ideas]
With Xcode 13 we have the option to use cloud signing. Basically if we `-allowProvisioningUpdates` we can now use the App Store api key instead of a developer id logged into Xcode to use automatic provisioning and sign without needing to install provisioning profiles from tools like `match`.

The steps are described in https://developer.apple.com/videos/play/wwdc2021/10204/.

These `xcodebuild` options are used to achieve this. 
`-allowProvisioningUpdates -authenticationKeyIssuerID <UUID> -authenticationKeyID <KEY_ID> -authenticationKeyPath <PATH_TO_KEY>` 

It would be nice to add support for the above in `gym`

**Comment by chkpnt** [comment]
Should already work using the general `xcargs` parameter.

But of course, it would be nice if the same arguments as [`upload_to_testflight`](https://docs.fastlane.tools/actions/upload_to_testflight/) were supported, i.e.
`--api_key_path ./path/to/api_key_info.json` or `--api_key "{\"key_id\": \"D83848D23\", \"issuer_id\": \"227b0bbf-ada8-458c-9d62-3d8022b7d07f\", \"key_filepath\": \"D83848D23.p8\"}"`

**Reply by tahirmt** [reply]
I kept getting code signing errors for adhoc even though it worked for TestFlight. That might be an Xcode issue.

**Reply by therohansanap** [reply]
@chkpnt Should we pass this in `xcargs` or `export_xcargs`? I am getting confused between these two not able to figure out which is correct place for this.

**Reply by pchmelar** [reply]
I'm using `xcargs`

**Comment by TSkovsgaard** [comment]
This is awesome, it works, but only the first time, I'm building with CircleCI, first time everything is smooth, the app is signed and uploaded to TestFlight, but the second time I get this error

`error: Revoke certificate: Your account already has an Apple Development signing certificate for this machine, but its private key is not installed in your keychain. Xcode can create a new one after revoking your existing certificate.`

I understand why this is happening, but what would be the best way to handle this? 

The perfect setup would be to not deal with profiles and certificates at all, but I might end up by still keeping the certificate handled by fastlane match. 

Or maybe just revoke the certificate after each build?

**Reply by tahirmt** [reply]
I think Xcode is supposed to manage it for you. I only tried with distribution and it always succeeded using the same certificate. It never revoked it.

**Reply by tahirmt** [reply]
The error message sounds to me like it’s using local signing and not cloud signing.

**Reply by TSkovsgaard** [reply]
The thing is, it works if I revoke the automatic created developer certificate and rebuild on circleCI. 
Another thing to note is it is the Apple Development certificate that is the issue, not the Managed Distribution certificate, or at least that's what the error says, the error happens before it actually builds the project right after _Summary for gym_, so I don't think it has anything to do with the final signing.
This is how I call it:

```
build_ios_app(
      clean: true,
      scheme: options[:scheme],
      project: 'Project.xcodeproj',
      output_name: 'app.ipa',
      include_symbols: true,
      export_method: "app-store",
      export_options: { 
        manageAppVersionAndBuildNumber: true 
      },
      xcargs: "-allowProvisioningUpdates -authenticationKeyIssuerID xxx -authenticationKeyID xxx -authenticationKeyPath #{File.expand_path('path_to_key')}"
    )

```

Steps leading up to the error:

```
[09:06:24]: ▸ Clean Succeeded
[09:06:25]: ▸ 2022-03-10 09:06:25.238 xcodebuild[2706:13925] CFURLRequestSetHTTPCookieStorageAcceptPolicy_block_invoke: no longer implemented and should not be called
[09:06:25]: ▸ 2022-03-10 09:06:25.669 xcodebuild[2706:13925]  DVTAssertions: Warning in /Library/Caches/com.apple.xbs/Sources/IDEFrameworks/IDEFrameworks-19586/IDEFoundation/Provisioning/Logging/IDEProvisioningLedger.m:172
[09:06:25]: ▸ Details:  Unable to close provisioning ledger entry because not all of its subentries are closed
[09:06:25]: ▸ Object:   <IDEProvisioningLedgerEntry: 0x7fba8f25f740>
[09:06:25]: ▸ Method:   -closeWithError:
[09:06:25]: ▸ Thread:   <NSThread: 0x7fba8fb164e0>{number = 68, name = (null)}
[09:06:25]: ▸ Please file a bug at https://feedbackassistant.apple.com with this warning message and any useful information you can provide.
[09:06:34]: ▸ ❌  error: Revoke certificate: Your account already has an Apple Development signing certificate for this machine, but its private key is not installed in your keychain. Xcode can create a new one after revoking your existing certificate.
```

**Reply by gcazangi** [reply]
@TSkovsgaard have you found a solution to this Apple Development signing certificate issue?

**Reply by TSkovsgaard** [reply]
Kind of, I created a developer certificate to be used for building on circleCI, then automatic signing works.

**Reply by Kingamattack** [reply]
@TSkovsgaard but if you use the uploaded developer certificate, does it make sense to have automatic signing? I have the exact same issue on Bitrise and my builds end up using the dev certificate instead of the one created by Xcode

**Reply by ju135** [reply]
@TSkovsgaard so you still need to use Match for code signing? Would be great if there was a way to get everything done just with the API key.

**Reply by TSkovsgaard** [reply]
Haven't looked more into this since I set it up with the dev certificate, it works and i think it's still a better solution than manually signing :)

**Reply by Ashton-W** [reply]
Xcode 26 still behaves this way.

**Comment by gcazangi** [comment]
Seems like the "Xcode cloud signing" works as expected only for managing the Distribution certificate when using multiple machines for building. You will not be able to get automatic handling of the Development certificate if you use different machines every run (as in the CI services). Just like before, you are still required to have the Development certificate + private key installed on the machine to be able to create an archive. After that, you may use the cloud signing parameters to actually export your archive to an IPA and sign it with the managed distribution certificate that is automatically handled by Xcode Cloud Signing.

You may split the process in two phases, this applies to Gym or any other tool calling xcodebuild directly:

1. Build the project and archive -- always demands a local **Development** certificate + private key installed on the building machine even if you provide authentication parameters to xcodebuild command, eg: 

```
xcodebuild archive \
-archivePath ~/Downloads/ \
-workspace Project.xcworkspace \
-scheme 'Project' \
-authenticationKeyIssuerID 123123123-6874-1233-3232-23123123123 \
-authenticationKeyID 123123123AAB \
-authenticationKeyPath ~/Downloads/AuthKey_123123123AAB.p8 \
-allowProvisioningUpdates \
-destination generic/platform=iOS
```

2. Export the archive to iPA and sign with **Distribution** certificate, handled by Xcode Cloud signing, eg:

```
xcodebuild -exportArchive \
-archivePath ~/Downloads/Project.xcarchive \
-exportOptionsPlist ~/Downloads/exportOptions.plist \
-exportPath ~/Downloads/ExportedProject.ipa \
-authenticationKeyIssuerID 123123123-6874-1233-3232-23123123123 \
-authenticationKeyID 123123123AAB \
-authenticationKeyPath ~/Downloads/AuthKey_123123123AAB.p8 \
-allowProvisioningUpdates
```

You can then finally push the IPA to Testflight using whatever you are used to.

If you just using a single machine to build and archive, then you don't have to manually handle the Development certificate + private key as that pair will be generated automatically on the first run and subsequent builds will simply make use of it.

**Comment by pchmelar** [comment]
This is amazing! 🤩 I'm in the agency business, and it always seemed like an overhead to set up `match` when developing a relatively small app for a client. Now we can check `Automatically manage signing` in the Xcode project, and everything works out of the box.

In the case of self-hosted CI, I'm using `get_certificates` to use a proper developer certificate (and create it if it doesn't exist). When using some cloud-based CI (with a new instance for every build), I'm using `import_certificate` to import a developer certificate stored in the repository. Xcode's cloud signing entirely manages distribution certificates and provisioning profiles.

It would be nice if `gym` can pick up the API key from `app_store_connect_api_key` and use it for those `authenticationKeyID` / `authenticationKeyIssuerID` / `authenticationKeyPath` xcargs. @joshdholtz What do you think? 🙂

**Reply by jc-paris** [reply]
Hello @joshdholtz 👋 

Could we expect gym to use the api key from [`app_store_connect_api_key`](http://docs.fastlane.tools/actions/app_store_connect_api_key/#app_store_connect_api_key) to manage Xcode cloud signing ? 

It's been a while since this issue was open and I didn't find a clear answer whether it's a plan to add support for that or if using `xcargs` will remain the preferred approach.

Thanks for your feedback !

**Comment by jbrodriguez** [comment]
hello all, i'm trying to get this to work, but i'm getting a `security: SecItemCopyMatching: The specified item could not be found in the keychain.` error.

i think this is because i don't have the private key that corresponds to `development.cer`

i'm running this on a github action, so i have to `create_keychain` and `import_certificate` each time.

so, i'm guessing that i need to actually do all the steps from the examples section of the [import_certificate docs](https://docs.fastlane.tools/actions/import_certificate/)

```
import_certificate(certificate_path: "certs/AppleWWDRCA6.cer")

import_certificate(
  certificate_path: "certs/dist.p12",
  certificate_password: ENV["CERTIFICATE_PASSWORD"] || "default"
)

import_certificate(
  certificate_path: "certs/development.cer"
)
```

this is my fastfile definition

```
  desc "Submit a new Beta Build to Apple TestFlight"
  lane :beta do
    create_keychain(
      name: "test.keychain-db",
      password: "test123",
      default_keychain: true,
      unlock: true,
      timeout: 3600,
      lock_when_sleeps: true
    )    
    
    import_certificate(
      certificate_path: ENV['APPLE_DEVELOPMENT_CERT'],
      keychain_name: "test.keychain-db",
      keychain_password: "test123"
    )

    key_path = ENV['APPSTORE_CONNECT_KEY_PATH']
    xcargs = "-allowProvisioningUpdates -authenticationKeyIssuerID <iuid> -authenticationKeyID <kid> -authenticationKeyPath #{key_path}"

    build_app(
      workspace: "ios/test.xcworkspace",
      scheme: "test",
      export_method: "app-store",
      xcargs: xcargs,
    )

    # upload_to_testflight
  end
```

can anoyone confirm ? @pchmelar maybe ? :)

p.s.: i've looked at the following troubleshooting docs
- https://developer.apple.com/forums/thread/701696
- https://developer.apple.com/forums/thread/699268

**Reply by pchmelar** [reply]
Hey 👋 Your configuration seems correct, `.p12` should be imported automatically when you import `.cer`
That error may be caused by incorrect WWDR authority, i.e. your certificate uses WWDR that is for some reason not available when you build the app on GitHub Action via fastlane. Related issues: https://github.com/fastlane/fastlane/issues/14379 / https://github.com/fastlane/fastlane/issues/20509

**Reply by jbrodriguez** [reply]
hi, thanks for the reply !

yes, it seems to be related to that

instead of creating the new keychain, i would like to use the default keychain, which i believe contains the WWDR certificate (whereas creating a new keychain may not), but i had some issues doing that

just one question, one can the .p12 be imported by the .cer ? i understand they're separate keys (p12/private, cer/public), so i think one (cer) shouldn't contain the other (p12) ?

**Reply by pchmelar** [reply]
I'm not 100% sure how it's implemented, but I had no problems with importing only `.p12`
The issue with the default keychain is that it's usually locked. I resolved the issue by explicitly importing the WWDR certificate before building the app. Something like this:

```ruby
    keychain_password = SecureRandom.base64
    create_keychain(
      name: "fastlane",
      unlock: true,
      timeout: 3600,
      lock_when_sleeps: true,
      password: keychain_password
    )
    import_certificate(
      certificate_path: "signing/AppleWWDRCAG3.cer",
      keychain_name: "fastlane",
      keychain_password: keychain_password
    )
    import_certificate(
      certificate_path: "signing/Development.p12",
      keychain_name: "fastlane",
      keychain_password: keychain_password
    )
    build_ios_app(
      ...
      xcargs: "-allowProvisioningUpdates -authenticationKeyID #{config[:api_key_id]} -authenticationKeyIssuerID #{config[:api_key_issuer_id]} -authenticationKeyPath #{File.expand_path('../AuthKey.p8')}"
    )

```

**Reply by jbrodriguez** [reply]
thanks a lot @pchmelar, this is very helpful !

**Comment by LyraHarring** [comment]
With Xcode 13, Apple added cloud code signing support so you don’t need to manage provisioning profiles manually. Using -allowProvisioningUpdates with your App Store Connect API key lets Xcode handle profiles and signing automatically. Right now gym doesn’t expose these flags directly, but if it did, it would make builds simpler and avoid the need for tools like match. Adding this support would be a great step forward for [cloud code signing](https://signmycode.com/cloud-code-signing).
